src/crates-io/lib.rs

```diff
@@ -300,55 +308,62 @@ impl Registry {
             Some(mut body) => {
                 self.handle.upload(true)?;
                 self.handle.in_filesize(body.len() as u64)?;
-                handle(&mut self.handle, &mut |buf| body.read(buf).unwrap_or(0))
+                self.handle(&mut |buf| body.read(buf).unwrap_or(0))
             }
-            None => handle(&mut self.handle, &mut |_| 0),
+            None => self.handle(&mut |_| 0),
         }
     }
-}
 
-fn handle(handle: &mut Easy, read: &mut dyn FnMut(&mut [u8]) -> usize) -> Result<String> {
-    let mut headers = Vec::new();
-    let mut body = Vec::new();
-    {
-        let mut handle = handle.transfer();
-        handle.read_function(|buf| Ok(read(buf)))?;
-        handle.write_function(|data| {
-            body.extend_from_slice(data);
-            Ok(data.len())
-        })?;
-        handle.header_function(|data| {
-            headers.push(String::from_utf8_lossy(data).into_owned());
-            true
-        })?;
-        handle.perform()?;
-    }
+    fn handle(&mut self, read: &mut dyn FnMut(&mut [u8]) -> usize) -> Result<String> {
+        let mut headers = Vec::new();
+        let mut body = Vec::new();
+        let started;
+        {
+            let mut handle = self.handle.transfer();
+            handle.read_function(|buf| Ok(read(buf)))?;
+            handle.write_function(|data| {
+                body.extend_from_slice(data);
+                Ok(data.len())
+            })?;
+            handle.header_function(|data| {
+                headers.push(String::from_utf8_lossy(data).into_owned());
+                true
+            })?;
+            started = Instant::now();
+            handle.perform()?;
+        }
 
-    let body = match String::from_utf8(body) {
-        Ok(body) => body,
-        Err(..) => bail!("response body was not valid utf-8"),
-    };
-    let errors = serde_json::from_str::<ApiErrorList>(&body).ok().map(|s| {
-        s.errors.into_iter().map(|s| s.detail).collect::<Vec<_>>()
-    });
-
-    match (handle.response_code()?, errors) {
-        (0, None) | (200, None) => {},
-        (code, Some(errors)) => {
-            let code = StatusCode::from_u16(code as _)?;
-            bail!("api errors (status {}): {}", code, errors.join(", "))
+        let body = match String::from_utf8(body) {
+            Ok(body) => body,
+            Err(..) => bail!("response body was not valid utf-8"),
+        };
+        let errors = serde_json::from_str::<ApiErrorList>(&body).ok().map(|s| {
+            s.errors.into_iter().map(|s| s.detail).collect::<Vec<_>>()
+        });
+
+        match (self.handle.response_code()?, errors) {
+            (0, None) | (200, None) => {},
+            (503, _) if started.elapsed().as_secs() >= 29 && self.host_is_crates_io() => {
```

Comment:
Could this perhaps be situated beneath the `(code, Some(errors))` case below? It seems like we would want to favor displaying any errors from crates.io first and then if nothing is listed try to provide our own flavortext?

Reply:
So in theory this branch could actually just be `(503, None)` since Heroku returns HTML (I'm not sure if it's ignoring the `Accept` header or if Cargo doesn't send one for publish). The reason I opted to give this higher precedent is that we don't actually control the response in the case of a timeout, and on the off chance Heroku starts giving us JSON with an errors key I think we'd still want to show this message instead.

"It took ~30 seconds to get a 503 back" is pretty clear that it's a timeout, so I figured that separated it enough. The only time we ever return a 503 is if we're in read-only mode. I've also seen folks get 503s from proxy issues. Both come back quickly.

That said, I don't feel strongly about it, so I'm happy to make that change if you still think we should

Reply:
This can also race though in scenarios with legitimate delayed error messages coming back from crates.io? For example if crates.io sends a legitimate error message around the 30s boundary or maybe something crazy like I suspend my laptop after I get errors from crates.io but before the time measurement here is taken (or something like that)

Reply:
Fair enough. Changed this to always log the errors given by the API if any were returned.